prdoc/pr_1378.prdoc

```diff
@@ -0,0 +1,27 @@
+title: Construct Runtime V2 - An outer macro approach to define the runtime
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      Introduces `#[frame_support::runtime]` that can be attached to a mod to define a runtime. The items 
+      in this mod can be attached to the following attributes to define the key components of the runtime.
+      1. `#[runtime::runtime]` attached to a struct defines the main runtime
+      2. `#[runtime::derive]` attached to the runtime struct defines the types generated by the runtime
+      3. `#[runtime::pallet_index]` must be attached to a pallet to define its index
+      4. `#[runtime::disable_call]` can be optionally attached to a pallet to disable its calls
+      5. `#[runtime::disable_unsigned]` can be optionally attached to a pallet to disable unsigned calls
+      6. A pallet instance can be defined as `TemplateModule: pallet_template<Instance>`
+      An optional attribute can be defined as `#[frame_support::runtime(legacy_ordering)]` to ensure that 
+      the order of hooks is same as the order of pallets (and not based on the pallet_index). This is to support 
+      legacy runtimes and should be avoided for new ones.
```

Comment:
@bkchr do you have an opinion on this?

Reply:
Hmm. Not sure this is such a great idea (ordering based on the index). But also wasn't the idea that this would be changed with the new `Poll` logic? Aka that the runtime implementor needs to explicitly state this order?

Reply:
> Hmm. Not sure this is such a great idea (ordering based on the index). 

I thought this was causing a lot of issues as the user expectation would be to have everything as per the `pallet_index`. I would also imagine this to be better but please let me know your concerns.

> But also wasn't the idea that this would be changed with the new Poll logic? Aka that the runtime implementor needs to explicitly state this order?

I am not aware of this. Could you please link the issue?

Reply:
For the poll hook it can explicitly be set, yes. But my question was about the `on_initialize` etc that have fixed order.

Reply:
> I thought this was causing a lot of issues as the user expectation would be to have everything as per the `pallet_index`

Yeah that is a good point :+1: 


> For the poll hook it can explicitly be set, yes. But my question was about the `on_initialize` etc that have fixed order.

If we have poll explicit and have everything fixed up with `on_initialize`, the order of `on_initialize` shouldn't be that important anymore. By "fixed up" `on_initialize` I mean that pallets should not call into other pallets in this stage if possible, to not run into race conditions that stuff is not yet setup. They should setup their own state and then in `poll` they can do whatever they need.